src/Core/Addon/Module/ModuleManager.php

```diff
@@ -386,12 +417,38 @@ public function upgrade($name, $version = 'latest', $source = null)
         $this->checkIsInstalled($name);
         $module = $this->moduleRepository->getModule($name);
 
-        // Get new module
-        // 1- From source
-        if ($source != null) {
-            $this->moduleZipManager->storeInModulesFolder($source);
+        if ($source === null) {
+            // Give a chance to the other modules (like marketplace) to retrieve the module source
+            $sourceZip = $this->hookManager->exec(
+                'actionAdminModuleUpgradeRetrieveSource',
+                [
+                    'name' => $name,
+                    'current_version' => $module->get('version'),
+                ],
+                null,
+                true
+            );
+
+            if (is_array($sourceZip)) {
+                $sourceZip = array_values($sourceZip)[0];
+            }
+
+            // We need to check that the file is actually a module's ZIP and the same module as the one expected
+            if (is_file($sourceZip)) {
```

Comment:
If strict mode was enabled in this PHP file, `is_file` would only accept a `string` as argument. Or trigger a PHP error otherwise.
If I was paranoid, I would add a quick validation first:
```suggestion
            if (is_string($sourceZip) && is_file($sourceZip)) {
```